crates/storage/provider/src/providers/mod.rs

```diff
@@ -0,0 +1,144 @@
+use crate::{BlockHashProvider, BlockProvider, Error, HeaderProvider, StateProviderFactory};
+use reth_db::{
+    database::{Database, DatabaseGAT},
+    tables,
+    transaction::DbTx,
+};
+use reth_interfaces::Result;
+use reth_primitives::{rpc::BlockId, Block, BlockHash, BlockNumber, ChainInfo, Header, H256, U256};
+use std::sync::Arc;
+
+mod historical;
+pub use historical::{HistoricalStateProvider, HistoricalStateProviderRef};
+
+mod latest;
+pub use latest::{LatestStateProvider, LatestStateProviderRef};
+
+/// A common provider that fetches data from a database.
+///
+/// This provider implements most provider or provider factory traits.
+pub struct DatabaseProvider<DB: Database> {
```

Comment:
I'm not sure about the naming here, because this is essentially just the Database.

The `*Provider` implementations all "provide" access to some data from the database.

I also think we'll eventually have a bunch of different Database wrappers (caching etc.., perhaps memory for testing)

Perhaps we just name this `Database`?

Reply:
i agree that it's not ideal, but `Database` name would conflict with `Database` trait. this is a tough one to name

Reply:
right, perhaps `ShareableDatabase`, since this is just an Arc wrapper.

This also means we could implement the providers entirely over `T: Database`, right?

Reply:
wdyt about getting rid of `DatabaseProvider` altogether?

Reply:
rn I think this is easier.

I guess if we need providers with a caching layer then we're still able to implement them, which is good, like

```rust
struct CachedWhatever<DB> {
  db: DB,
  lru
}

impl SomeProvider for CachedWhatever {...}

```